src/wlapplication_options.cc

```diff
@@ -113,6 +119,212 @@ void set_config_string(const std::string& section,
 	g_options.pull_section(section.c_str()).set_string(name.c_str(), value.c_str());
 }
 
+static std::map<KeyboardShortcut, SDL_Keysym> shortcuts_;
+static void write_shortcut(const KeyboardShortcut id, const SDL_Keysym code) {
+	set_config_int("keyboard", std::to_string(static_cast<int>(id)) + "sym", code.sym);
+	set_config_int("keyboard", std::to_string(static_cast<int>(id)) + "mod", code.mod);
+}
+void set_shortcut(const KeyboardShortcut id, const SDL_Keysym code) {
+	shortcuts_.at(id) = code;
+	write_shortcut(id, code);
+}
+
+SDL_Keysym get_shortcut(const KeyboardShortcut id) {
+	return shortcuts_.at(id);
+}
+
+bool matches_shortcut(const KeyboardShortcut id, const SDL_Keysym code) {
+	const SDL_Keysym key = get_shortcut(id);
+	if (key.sym != code.sym) {
+		return false;
+	}
+
+	const bool ctrl1 = code.mod & KMOD_CTRL;
+	const bool shift1 = code.mod & KMOD_SHIFT;
+	const bool alt1 = code.mod & KMOD_ALT;
+	const bool ctrl2 = code.mod & KMOD_CTRL;
+	const bool shift2 = code.mod & KMOD_SHIFT;
+	const bool alt2 = code.mod & KMOD_ALT;
+
+	return ctrl1 == ctrl2 && shift1 == shift2 && alt1 == alt2;
+}
+
+std::string shortcut_string_for(const KeyboardShortcut id) {
+	const SDL_Keysym sym = get_shortcut(id);
+
+	std::vector<std::string> mods;
+	if (sym.mod & KMOD_CTRL) {
+		mods.push_back(pgettext("hotkey", "Ctrl"));
+	}
+	if (sym.mod & KMOD_SHIFT) {
+		mods.push_back(pgettext("hotkey", "Shift"));
+	}
+	if (sym.mod & KMOD_ALT) {
+		mods.push_back(pgettext("hotkey", "Alt"));
+	}
+
+	std::string fmt;
+	switch (mods.size()) {
+	case 0:
+		fmt = _("%1$s");
+		break;
+	case 1:
+		fmt = _("%1$s+%2$s");
+		break;
+	case 2:
+		fmt = _("%1$s+%2$s+%3$s");
+		break;
+	case 3:
+		fmt = _("%1$s+%2$s+%3$s+%4$s");
+		break;
+	default:
+		NEVER_HERE();
+	}
+	boost::format f(fmt);
+	for (const std::string& m : mods) {
+		f % m;
+	}
+	f % SDL_GetKeyName(sym.sym);
+	return f.str();
+}
+
+std::string to_string(const KeyboardShortcut id) {
+	switch (id) {
+	case KeyboardShortcut::kMainMenuNew:
```

Comment:
Can we create a struct for the names & defaults and stick them in a map? I think that would be easier to read, because we'll have all the definitions for a shortcut next to each other.

They will also need an internal name so that we can expose them to the Lua interface for the Encyclopedia, game tips & Tutorials.